src/NodeUtils.jl

```diff
@@ -69,19 +71,46 @@ whether it depends on input features.
 is_constant(tree::AbstractExpressionNode) = all(t -> t.degree != 0 || t.constant, tree)
 
 """
-    get_constants(tree::AbstractExpressionNode{T})::Vector{T} where {T}
+    count_number_constants(tree::AbstractExpressionNode{T})::Int64 where {T}
 
-Get all the constants inside a tree, in depth-first order.
+Returns the number of number constants in the tree.
+Used in get_constants to preallocate the constants array.
+"""
+function count_number_constants(tree::AbstractExpressionNode{T}) where {T}
+    return tree_mapreduce(
+        node -> is_node_constant(node) ? count_number_constants(node.val) : 0,
+        +,
+        tree,
+        Int64;
+        f_on_shared=(c, is_shared) -> is_shared ? 0 : c,
+    )
+end
+
+"""
+    get_constants(tree::AbstractExpressionNode{T}, BT::Type = T)::Vector{T} where {T}
+
+Get all the number constants inside a tree, in depth-first order.
 The function `set_constants!` sets them in the same order,
 given the output of this function.
 Also return metadata that can will be used in the `set_constants!` function.
 """
-function get_constants(tree::AbstractExpressionNode{T}) where {T}
+function get_constants(
+    tree::AbstractExpressionNode{T}, BT::Type=get_number_type(T)
+) where {T}
     refs = filter_map(
         is_node_constant, node -> Ref(node), tree, Base.RefValue{typeof(tree)}
     )
-    return map(ref -> ref[].val::T, refs), refs
-    # NOTE: Do not remove this `::T` as it is required for inference on empty collections
+    if T <: Number
+        # NOTE: Do not remove this `::T` as it is required for inference on empty collections
+        return map(r -> r[].val::T, refs), refs
+    else
+        vals = BT[]
+        sizehint!(vals, count_number_constants(tree))
+        for i in eachindex(refs)
+            append_number_constants!(vals, refs[i][].val::T)
+        end
+        return vals, refs
```

Comment:
```suggestion
        vals = Vector{BT}(undef, count_number_constants(tree))
        i = firstindex(vals)
        for ref in refs
            i = append_number_constants!(vals, i, ref[].val::T)
        end
        return vals, refs
```

This might be a faster way? So, basically, `append_number_constants!` would be something like

```julia
start = i
stop = i + num_constants_per_ref
vals[start:stop] = ...
return stop + 1
```

It would return `stop + 1` so that the next call to `append_number_constants!` knows what index of `vals` to start at.